### DISCUSSION TITLE
How to raise an error?

### DISCUSSION THREAD

**Discussion by ales-tsurko** [koto-lang/koto#54] [Q&A] [answered]
I see try/catch, but can't find how to actually raise an error. Is it possible?

**Comment by irh** [comment]
It's not possible at the moment, other than using assertions or deliberately doing something invalid.

IIRC I was about to work on `throw` after adding `try`/`catch`, but then I moved on to other things and it slipped my mind.

As things stand `throw` would only support strings being thrown, but I would like to support custom error types and I haven't got a complete picture of what that would look like yet.

**Reply by irh** [reply]
I think what could work is that `throw` would accept a string, or any map that implements `@display` (`@display` doesn't exist yet, but it would be useful in general for overriding string formatting behaviour).

```coffee
try
  throw 
    data: foo
    @display: |self| "A terrible error occurred: {}".format self.data
catch error
  log_error_data error.data
  throw error
```

...then if the error is returned from the runtime then a meaningful error message would be available.

**Reply by ales-tsurko** [reply]
Using maps would also allow making different types of errors.

**Reply by irh** [reply]
Yes exactly, you can stick any data you like in the map, and it can have any structure as long as `@display` is implemented.

**Reply by irh** [reply]
I've implemented `@display` in #55, and I'll open an issue for `throw`.

**Reply by irh** [reply]
Issue for `throw`: #56.

**Comment by irh** [comment, marked as answer]
`throw` is now available, added in #57.
